### PR TITLE
feat(ast) Add a uuid optimizer for conditions with UUID columns

### PR DESCRIPTION
**Pull request by evanh**
First pass at trying to optimize queries with UUID columns. Right now the simple
case is to target conditions that contain UUID columns, and do a direct
comparison instead of converting the column to a string.

### CODE REVIEW

File: snuba/datasets/storages/transactions.py [hunk 1]
```diff
@@ -91,6 +92,7 @@
         EventIdColumnProcessor(),
```
Comment (lynnagara) [inline snuba/datasets/storages/transactions.py:92]: Isn't this the one we should be replacing?

File: snuba/query/processors/uuid_column_processor.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+from snuba.query.matchers import (
+    AnyOptionalString,
+    Or,
+    Param,
+    String,
+)
+from snuba.query.matchers import Column as ColumnMatch
+from snuba.query.matchers import FunctionCall as FunctionCallMatch
+from snuba.query.matchers import Literal as LiteralMatch
+from snuba.query.processors import QueryProcessor
```
Comment (lynnagara) [inline snuba/query/processors/uuid_column_processor.py:20]: I think this needs to be the Clickhouse QueryProcessor, not the logical query one.

File: snuba/query/processors/uuid_column_processor.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            else:
+                column = result.expression("uuid_column")
+                assert isinstance(column, Column)
+                new_params.append(column)
+
+            params_fn = result.expression("params")
+            assert isinstance(params_fn, FunctionCall)
+            new_fn_params = []
+            for param in params_fn.parameters:
+                new_fn_params.append(self.parse_uuid(param))
```
Comment (fpacifici) [inline snuba/query/processors/uuid_column_processor.py:114]: Should we bail entirely and not transform the expression entirely if we see something in the parameters that is not a Literal. That parameter could be a string function around the UUID litral. 
If we append that FunctionCall to new_fn_params we would make a valid query invalid. We would be transforming the left part of the  condition from `replaceAll(toString(event_id....) IN tuple(f('a_uuid'))` in `event_id IN tuple(f('a_uuid')`

File: snuba/query/processors/uuid_column_processor.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+            new_params.append(
+                FunctionCall(
+                    params_fn.alias, params_fn.function_name, tuple(new_fn_params)
+                )
+            )
+
+            return binary_condition(
+                exp.alias, exp.function_name, new_params[0], new_params[1]
```
Comment (fpacifici) [inline snuba/query/processors/uuid_column_processor.py:123]: new_params is always a sequence of two elements that you are exploding here anyway. It would be more readable if you just gave a name to each and avoid the list. Then by reading this statement you would easily know what is new_param[0] and what is new_param[1] without trying to understand why this is a list.

File: snuba/query/processors/uuid_column_processor.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+                    new_params.append(result.expression("uuid_column" + suffix))
+                elif result.contains("formatted_uuid_column" + suffix):
+                    column = result.expression("formatted_uuid_column" + suffix)
+                    assert isinstance(column, Column)
+                    new_params.append(
+                        Column(None, column.table_name, column.column_name)
+                    )
+
+            return binary_condition(
+                exp.alias, exp.function_name, new_params[0], new_params[1]
```
Comment (fpacifici) [inline snuba/query/processors/uuid_column_processor.py:144]: same comment about new_params I made above.

File: snuba/query/processors/uuid_column_processor.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        if result is not None:
+            new_params: List[Expression] = []
+            if result.contains("formatted_uuid_column"):
+                column = result.expression("formatted_uuid_column")
+                assert isinstance(column, Column)
+                new_params.append(Column(None, column.table_name, column.column_name))
+            else:
+                column = result.expression("uuid_column")
+                assert isinstance(column, Column)
+                new_params.append(column)
```
Comment (fpacifici) [inline snuba/query/processors/uuid_column_processor.py:108]: I don't think we need to deal with this case.
If I get it right this identifies `event_id IN tuple('a7d67cf796774551a95be6543cacd459')` to transform it in `event_id IN tuple('a7d67cf7-9677-4551-a95b-e6543cacd459')`.
That to me seems just an invalid query we should reject it instead of trying to fix it as it is a type error.

I would just skip that case, it would simplify the logic of this processor, which is already quite complex.

**Review by fpacifici — COMMENTED** [review]
A couple of comments inline that I think would make this easier to follow otherwise I don't see any issue.

File: snuba/datasets/storages/transactions.py (hunk 1 above)
Comment (evanh) [reply snuba/datasets/storages/transactions.py:92]: I didn't realize that logic had been moved to a processor. However it's not that simple, because we still want the EventId processor to run, so the return values to Sentry are correct. The UUID processor I added is only used in the condition clauses.

File: snuba/query/processors/uuid_column_processor.py (hunk 3 above)
Comment (evanh) [reply snuba/query/processors/uuid_column_processor.py:114]: Yeah, that makes sense. I'll change it.

File: snuba/query/processors/uuid_column_processor.py (hunk 6 above)
Comment (evanh) [reply snuba/query/processors/uuid_column_processor.py:108]: Why is that an invalid query? It seems like a very possible use case to have IN conditions on event_ids. Why not optimize that case?

File: snuba/datasets/storages/transactions.py (hunk 1 above)
Comment (lynnagara) [reply snuba/datasets/storages/transactions.py:92]: I see. We are never actually formatting `trace_id` like `event_id` though (perhaps we should be). Does this still need to be handled in the UUID column processor then? The treatment of these columns at the moment is quite inconsistent.

Comment (lynnagara) [reply snuba/datasets/storages/transactions.py:92]: I think ideally the UUID column processor would handle the formatting in all parts of the query, rather than rely on the user adding two separate processors (the one that handles the conditions part and the one that handles the other parts of the query) in the correct order and on the same columns which seems easy to get wrong.

File: snuba/query/processors/uuid_column_processor.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+                        self.parse_uuid(result.expression("literal" + suffix))
+                    )
+                elif result.contains("uuid_column" + suffix):
+                    new_params.append(result.expression("uuid_column" + suffix))
+                elif result.contains("formatted_uuid_column" + suffix):
+                    column = result.expression("formatted_uuid_column" + suffix)
+                    assert isinstance(column, Column)
+                    new_params.append(
+                        Column(None, column.table_name, column.column_name)
+                    )
```
Comment (lynnagara) [inline snuba/query/processors/uuid_column_processor.py:141]: Is this just equivalent to `new_params.append(column)`?

**Comment by lynnagara** [comment]
Is it worth adding a test on `trace_id` to `TestTransactionsApi` - it's unclear to me what the expected behaviour on that field is

File: snuba/query/processors/uuid_column_processor.py (hunk 6 above)
Comment (fpacifici) [reply snuba/query/processors/uuid_column_processor.py]: I am talking about the transformation between `a7d67cf796774551a95be6543cacd459` into `a7d67cf7-9677-4551-a95b-e6543cacd459` when the original condition had not been processed into `replaceAll(toString.....))`. 
If, at this point of the query processing, we receive `event_id IN tuple('a7d67cf796774551a95be6543cacd459')` regarding an event_id column that is of type UUID, that query is invalid (that condition, if it was a valid case, should have been transformed earlier so either we failed in the transformation before or the query was invalid to begin with).

I would not fix an invalid query turning it into a valid one in a processor that is supposed to deal with performance only. Otherwise we silently start depending on this processor to be executed here for the correctness of the queries as it covers either a bug in the event id processor or in the translator. Or invalid queries from the client.

File: snuba/datasets/storages/transactions.py (hunk 1 above)
Comment (evanh) [reply snuba/datasets/storages/transactions.py:92]: I can make that change, but it will make this processor even more complicated. Do you think that's worth doing as part of this PR or should I make a separate task for it?

File: snuba/query/processors/uuid_column_processor.py (hunk 6 above)
Comment (evanh) [reply snuba/query/processors/uuid_column_processor.py]: Ah, I see your point here. OK I'll change this.

File: snuba/datasets/storages/transactions.py (hunk 1 above)
Comment (lynnagara) [reply snuba/datasets/storages/transactions.py:92]: Up to you, whichever is easier. I'd suggest to leave a TODO at least though.

**Comment by evanh** [comment]
> Is it worth adding a test on `trace_id` to `TestTransactionsApi` - it's unclear to me what the expected behaviour on that field is

@lynnagara I added a test for this and it made me realize that trace ID is being completely mishandled. Sentry is sending queries against trace id as `contexts[trace.trace_id]`, and then we are not using the promoted column at all in Snuba. So conditions look like `equals(_snuba_contexts[trace.trace_id], '69b1606fd5024665a904ec86cd054130')`. It's extra confusing because the context value is stored without the dashes so this condition works. If we change things so we use the promoted column instead of the context, then we will need to an EventIdColumnProcessor equivalent for trace_id and this processor will start optimizing those queries as well.

File: snuba/query/processors/uuid_column_processor.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        self.__uuid_column_match = Or([String(u_col) for u_col in uuid_columns])
+        self.uuid_in_condition = FunctionCallMatch(
+            Or((String(ConditionFunctions.IN), String(ConditionFunctions.NOT_IN))),
+            (
+                self.formatted_uuid_pattern(),
+                Param("params", FunctionCallMatch(String("tuple"), None)),
+            ),
+        )
+        self.uuid_condition = FunctionCallMatch(
+            Or([String(op) for op in FUNCTION_TO_OPERATOR]),
```
Comment (lynnagara) [inline snuba/query/processors/uuid_column_processor.py:58]: This includes the IN and NOT IN operators - intentional?

**Comment by fpacifici** [comment]
re: trace_id. I am pretty sure we were discussing about how to promote those context that are not strings months ago on this thread https://github.com/getsentry/snuba/pull/955#issuecomment-634191845.

The problem about the trace_id context is tracked on our backlog.

tl;dr: The processor is only part of the fix. To remove the issue entirely we need to address the context promotion as well.

**Review by fpacifici — COMMENTED** [review]
Looks good to me. @lynnagara ?

**Review by lynnagara — APPROVED** [review]
Looks good to me
